File: src/infrastructure/storage/markdown_storage_repository.py
from src.domain.entities.board import Board
from src.domain.entities.column import Column
from src.domain.entities.item import Item
from src.domain.repositories.storage_repository import StorageRepository
from src.utils.path_resolver import PathResolver
from src.utils.logger_factory import ContextAwareLogger
from src.infrastructure.storage.board_persistence import BoardPersistence
from src.infrastructure.storage.file_operations import find_item_file_by_id


class MarkdownStorageRepository(StorageRepository):
    def __init__(self, path_resolver: PathResolver, logger: ContextAwareLogger):
        self.path_resolver = path_resolver
        self.logger = logger
        self.persistence = BoardPersistence(path_resolver.get_data_dir())

    def delete_item_from_column(self, board: Board, item: Item, column: Column) -> bool:
        self.logger.debug(
            "Deleting item from column storage",
            board=board.name,
            column=column.name,
            item=item.title,
        )

        column_dir = self.path_resolver.get_column_directory(board.name, column.name)
        item_file = find_item_file_by_id(column_dir, item.id)

        if item_file and item_file.exists():
            try:
                item_file.unlink()
                self.logger.info(
                    "Successfully deleted item file",
                    board=board.name,
                    column=column.name,
                    item=item.title,
                )
                return True
            except Exception:
                self.logger.error(
                    "Failed to delete item file",
                    board=board.name,
                    column=column.name,
                    item=item.title,
                )
                return False

        self.logger.warning(
            "Item file not found for deletion",
            board=board.name,
            column=column.name,
            item=item.title,
        )
        return False

    def move_item_between_columns(
        self, board: Board, item: Item, old_column: Column, new_column: Column
    ) -> bool:
        self.logger.info(
            "Moving item between columns",
            board=board.name,
            item=item.title,
            column=f"{old_column.name} -> {new_column.name}",
        )

        # First save the item to the new column
        try:
            self.persistence.save_item_to_column(
                board.name,
                new_column.name,
                {
                    "id": item.id,
                    "title": item.title,
                    "description": item.description,
                    "parent_id": item.parent_id,
                    "created_at": item.created_at,
                    "moved_in_progress_at": item.moved_in_progress_at,
                    "moved_in_done_at": item.moved_in_done_at,
                    "worked_on_for": item.worked_on_for,
                },
            )

            self.logger.debug(
                "Saved item to new column",
                board=board.name,
                column=new_column.name,
                item=item.title,
            )

            # Then delete from the old column
            if self.delete_item_from_column(board, item, old_column):
                self.logger.info(
                    "Successfully moved item between columns",
                    board=board.name,
                    item=item.title,
                    column=f"{old_column.name} -> {new_column.name}",
                )
                return True
            else:
                # If delete failed, try to clean up the new file
                self.logger.warning(
                    "Failed to delete from old column, cleaning up new column",
                    board=board.name,
                    item=item.title,
                )
                self.delete_item_from_column(board, item, new_column)
                return False

        except Exception:
            self.logger.error(
                "Failed to save item to new column",
                board=board.name,
                column=new_column.name,
                item=item.title,
            )
            return False

    def save_board_to_storage(self, board: Board) -> None:
        self.logger.debug("Saving board to storage", board=board.name)

        try:
            # Save all columns and their items
            for column in board.columns:
                self.logger.debug("Saving column", board=board.name, column=column.name)

                # Ensure column directory exists
                self.path_resolver.get_column_directory(
                    board.name, column.name
                )

                # Save column metadata
                self.persistence.save_column_metadata(
                    board.name,
                    column.name,
                    {
                        "id": column.id,
                        "name": column.name,
                        "position": column.position,
                        "limit": column.limit,
                        "created_at": column.created_at,
                    },
                )

                # Save all items in this column
                for item in column.items:
                    self.persistence.save_item_to_column(
                        board.name,
                        column.name,
                        {
                            "id": item.id,
                            "title": item.title,
                            "description": item.description,
                            "parent_id": item.parent_id,
                            "created_at": item.created_at,
                            "moved_in_progress_at": item.moved_in_progress_at,
                            "moved_in_done_at": item.moved_in_done_at,
                            "worked_on_for": item.worked_on_for,
                        },
                    )

            self.logger.info("Successfully saved board to storage", board=board.name)

        except Exception:
            self.logger.error("Failed to save board to storage", board=board.name)
            raise

File: mobile/src/infrastructure/storage/MarkdownStorageRepository.ts
/**
 * Markdown Storage Repository
 * Item-level storage operations with rollback support
 * Ported from Python: src/infrastructure/storage/markdown_storage_repository.py
 */

import { FileSystemManager } from "./FileSystemManager";
import { MarkdownParser } from "./MarkdownParser";
import { BoardPersistence } from "./BoardPersistence";
import { findItemFileById } from "./FileOperations";
import { StorageRepository } from "../../domain/repositories/StorageRepository";
import { Board } from "../../domain/entities/Board";
import { Column } from "../../domain/entities/Column";
import { Item } from "../../domain/entities/Item";

export class MarkdownStorageRepository implements StorageRepository {
  private fileSystem: FileSystemManager;
  private parser: MarkdownParser;
  private persistence: BoardPersistence;

  constructor(fileSystem: FileSystemManager) {
    this.fileSystem = fileSystem;
    this.parser = new MarkdownParser(fileSystem);
    this.persistence = new BoardPersistence(fileSystem, this.parser);
  }

  /**
   * Delete an item from a column
   */
  async deleteItemFromColumn(board: Board, item: Item, column: Column): Promise<boolean> {
    try {
      console.log(
        `Deleting item from column: board="${board.name}", column="${column.name}", item="${item.title}"`
      );

      const columnDir = this.fileSystem.getColumnDirectory(board.name, column.name);
      const itemFile = await findItemFileById(
        this.fileSystem,
        this.parser,
        columnDir,
        item.id
      );

      if (itemFile) {
        const deleted = await this.fileSystem.deleteFile(itemFile);

        if (deleted) {
          console.log(
            `Successfully deleted item: board="${board.name}", column="${column.name}", item="${item.title}"`
          );
          return true;
        } else {
          console.error(
            `Failed to delete item file: board="${board.name}", column="${column.name}", item="${item.title}"`
          );
          return false;
        }
      }

      console.warn(
        `Item file not found for deletion: board="${board.name}", column="${column.name}", item="${item.title}"`
      );
      return false;
    } catch (error) {
      console.error(
        `Failed to delete item from column: board="${board.name}", column="${column.name}", item="${item.title}"`,
        error
      );
      return false;
    }
  }

  /**
   * Move an item between columns with rollback on failure
   */
  async moveItemBetweenColumns(
    board: Board,
    item: Item,
    oldColumn: Column,
    newColumn: Column
  ): Promise<boolean> {
    try {
      console.log(
        `Moving item between columns: board="${board.name}", item="${item.title}", from="${oldColumn.name}" to="${newColumn.name}"`
      );

      // First, save the item to the new column
      try {
        await this.persistence.saveItemToColumn(board.name, newColumn.name, {
          id: item.id,
          title: item.title,
          description: item.description,
          parent_id: item.parent_id,
          created_at: item.created_at,
          moved_in_progress_at: item.moved_in_progress_at,
          moved_in_done_at: item.moved_in_done_at,
          worked_on_for: item.worked_on_for,
        });

        console.log(
          `Saved item to new column: board="${board.name}", column="${newColumn.name}", item="${item.title}"`
        );

        // Then, delete from the old column
        const deleted = await this.deleteItemFromColumn(board, item, oldColumn);

        if (deleted) {
          console.log(
            `Successfully moved item between columns: board="${board.name}", item="${item.title}", from="${oldColumn.name}" to="${newColumn.name}"`
          );
          return true;
        } else {
          // If delete failed, try to clean up the new file (rollback)
          console.warn(
            `Failed to delete from old column, cleaning up new column: board="${board.name}", item="${item.title}"`
          );
          await this.deleteItemFromColumn(board, item, newColumn);
          return false;
        }
      } catch (saveError) {
        console.error(
          `Failed to save item to new column: board="${board.name}", column="${newColumn.name}", item="${item.title}"`,
          saveError
        );
        return false;
      }
    } catch (error) {
      console.error(
        `Failed to move item between columns: board="${board.name}", item="${item.title}"`,
        error
      );
      return false;
    }
  }

  /**
   * Save entire board to storage (all columns and items)
   */
  async saveBoardToStorage(board: Board): Promise<void> {
    try {
      console.log(`Saving board to storage: board="${board.name}"`);

      // Save all columns and their items
      for (const column of board.columns) {
        console.log(`Saving column: board="${board.name}", column="${column.name}"`);

        // Ensure column directory exists
        const columnDir = this.fileSystem.getColumnDirectory(board.name, column.name);
        await this.fileSystem.ensureDirectoryExists(columnDir);

        // Save column metadata
        await this.persistence.saveColumnMetadata(board.name, column.name, {
          id: column.id,
          name: column.name,
          position: column.position,
          limit: column.limit,
          created_at: column.created_at,
        });

        // Save all items in this column
        for (const item of column.items) {
          await this.persistence.saveItemToColumn(board.name, column.name, {
            id: item.id,
            title: item.title,
            description: item.description,
            parent_id: item.parent_id,
            created_at: item.created_at,
            moved_in_progress_at: item.moved_in_progress_at,
            moved_in_done_at: item.moved_in_done_at,
            worked_on_for: item.worked_on_for,
          });
        }
      }

      console.log(`Successfully saved board to storage: board="${board.name}"`);
    } catch (error) {
      console.error(`Failed to save board to storage: board="${board.name}"`, error);
      throw error;
    }
  }
}
